daffodil-japi/src/main/scala/org/apache/daffodil/japi/schema/RuntimeSchemaWalker.scala

```diff
@@ -0,0 +1,135 @@
+package org.apache.daffodil.japi.schema
+
+import org.apache.daffodil.japi.{ DataProcessor => JDataProcessor }
+import org.apache.daffodil.lib.exceptions.Assert
+import org.apache.daffodil.runtime1.dpath.NodeInfo.PrimType
+import org.apache.daffodil.runtime1.processors.{
+  ChoiceRuntimeData,
+  DataProcessor => SDataProcessor,
+  ElementRuntimeData,
+  ErrorERD,
+  SequenceRuntimeData,
+  TermRuntimeData,
+}
+
+/**
+ * Base class used by japi clients who want to walk the runtime schema information.
+ */
+abstract class RuntimeSchemaHandler() {
+
+  /**
+   * Called for simple type element declarations.
+   *
+   * @param qnameString
+   * @param primTypeName
+   * @param isArray only true if more than one occurrence is possible. Disjoint with isOptional
+   * @param isOptional true if 0 or 1 occurrence only. Allows for a different representation of optional from a 0..1 array
+   */
+  def elementSimple(
+    qnameString: String,
+    primTypeName: String,
+    isArray: Boolean,
+    isOptional: Boolean,
+  ): Unit
+
+  /**
+   * Called for complex type element declarations
+   *
+   * Subsequent calls will be for the model group making up the content
+   * of the element.
+   *
+   * @param qnameString
+   */
+  def startElementComplex(qnameString: String, isArray: Boolean, isOptional: Boolean): Unit
+  def endElementComplex(qnameString: String, isArray: Boolean, isOptional: Boolean): Unit
+
+  def startSequence(): Unit
+
+  def endSequence(): Unit
+
+  def startChoice(): Unit
+
+  def endChoice(): Unit
+
+}
+
+/**
+ * Walks the schema, but not the DSOM schema, it walks the RuntimeData objects that
+ * represent the DFDL schema at runtime.
+ *
+ * @param dp
+ */
+class RuntimeSchemaWalker(private val dp: SDataProcessor) {
+
+  // provided so that people can write various tests from JAPI only.
+  // we wouldn't need this otherwise.
+  def this(jdp: JDataProcessor) = this(jdp.getUnderlyingDataProcessor)
+
+  private lazy val rootERD = dp.ssrd.elementRuntimeData
+
+  def walk(handler: RuntimeSchemaHandler): Unit = {
+    walkTerm(handler, rootERD)
+  }
+
+  private def walkTerm(handler: RuntimeSchemaHandler, trd: TermRuntimeData): Unit = {
```

Comment:
I'm not sure where the logic should be, but one difference between this walker and the DSOM walker is the DSOM walker has a `isHidden` boolean to avoid walking into hidden group refs that won't end up in the infoset. Does this need something similar, or beause this is walking runtime1 constructs are hidden elements implicitly skipped over?

Reply:
I will have to verify the RuntimeSchemaWalker (or whatever we end up calling it) does not walk into hidden groups. We don't need the hidden groups because the infoset outputter from Daffodil via JAPI already has them removed from it. So we never will encounter an actual element for conversion to Drill that is hidden.